src/Application/Kernel.php

```diff
@@ -161,4 +149,57 @@ private function initDiscovery(Container $container): void
             $discovery->storeCache();
         }
     }
+
+    private function discoverInstalledPackageLocations(): void
+    {
+        $composerPath = path($this->root, 'vendor/composer');
+        $installed = path($composerPath, 'installed.json');
+
+        if (is_file($installed)) {
+            $installed = $this->loadJsonFile($installed);
+
+            foreach ($installed['packages'] ?? [] as $package) {
+                $packagePath = realpath(path($composerPath, $package['install-path']));
+
+                if (isset($package['require']['tempest/framework'], $package['autoload']['psr-4'])
+                    && is_array($package['autoload']['psr-4'])
+                ) {
+                    foreach ($package['autoload']['psr-4'] as $namespace => $namespacePath) {
+                        $namespacePath = path($packagePath, $namespacePath);
+                        $this->addDiscoveryLocation($namespace, $namespacePath);
+                    }
+                }
+            }
+        }
+    }
+
+    private function discoverTempestNamespaces(): void
+    {
+        $composer = $this->loadJsonFile(path($this->root, 'composer.json'));
+
+        $autoloads = [$composer['autoload']['psr-4'] ?? []];
```

Comment:
I don't understand what this part of the code does? Why the check on environment? What's the difference between these lines?

```
$autoloads = [$composer['autoload']['psr-4'] ?? []];

// …

$autoloads[] = $composer['autoload']['psr-4'] ?? [];
```

Reply:
Oops. i left out part of the key name inside the `if`. Should look like:

```php
if ( /* not production */ ) {
    $autoloads[] = $composer['autoload-dev']['psr-4'] ?? [];
}
```

We always want the production `"autoload"` set of PSR-4 namespaces. But the original code was also including the framework's `"Tests\\Tempest\\": "tests/"` namespace, which is in `"autoload-dev"`.

I didn't want to include dev namespaces in a production environment, but to keep continuity with the original code, this adds the `"autoload-dev"` section for non-production.

Reply:
I don't think we should take that into account here, we could manually add items to appconfig during test setup

Reply:
i'll cover this in your new question below.